x-pack/plugins/security_solution/server/lib/telemetry/endpoint_task.ts

```diff
@@ -0,0 +1,202 @@
+
+          return {
+            run: async () => {
+              const executeTo = moment().utc().toISOString();
+              const lastExecutionTimestamp = getLastTaskExecutionTimestamp(
+                executeTo,
+                taskInstance.state?.lastExecutionTimestamp
+              );
+
+              const hits = await this.runTask(taskInstance.id);
+
+              return {
+                state: {
+                  lastExecutionTimestamp,
+                  runs: (state.runs || 0) + 1,
+                  hits,
+                },
+              };
+            },
+            cancel: async () => {},
+          };
+        },
+      },
+    });
+  }
+
+  public start = async (taskManager: TaskManagerStartContract) => {
+    try {
+      await taskManager.ensureScheduled({
+        id: this.getTaskId(),
+        taskType: TelemetryEndpointTaskConstants.TYPE,
+        scope: ['securitySolution'],
+        schedule: {
+          interval: TelemetryEndpointTaskConstants.INTERVAL,
+        },
+        state: { runs: 0 },
+        params: { version: TelemetryEndpointTaskConstants.VERSION },
+      });
+    } catch (e) {
+      this.logger.error(`Error scheduling task, received ${e.message}`);
+    }
+  };
+
+  private getTaskId = (): string => {
+    return `${TelemetryEndpointTaskConstants.TYPE}:${TelemetryEndpointTaskConstants.VERSION}`;
+  };
+
+  public runTask = async (taskId: string) => {
+    if (taskId !== this.getTaskId()) {
+      this.logger.debug(`Outdated task running: ${taskId}`);
+      return 0;
+    }
+
+    const isOptedIn = await this.sender.isTelemetryOptedIn();
+    if (!isOptedIn) {
+      this.logger.debug(`Telemetry is not opted-in.`);
+      return 0;
+    }
+
+    const {
+      body: endpointMetricsResponse,
+    } = ((await this.sender.fetchEndpointMetrics()) as unknown) as {
+      body: EndpointMetricsAggregation;
+    };
+    const endpointMetrics = endpointMetricsResponse.aggregations.endpoint_agents.buckets.map(
+      (epMetrics) => {
+        return {
+          endpoint_agent: epMetrics.latest_metrics.hits.hits[0]._source.agent.id,
+          endpoint_metrics: epMetrics.latest_metrics.hits.hits[0]._source,
+        };
+      }
+    );
+
+    if (endpointMetrics.length === 0) {
+      this.logger.debug('no reported endpoint metrics');
+      return 0;
+    }
+
+    const agentsResponse = await this.sender.fetchFleetAgents();
+    if (agentsResponse === undefined) {
+      this.logger.debug('no agents to report');
+      return 0;
+    }
+
+    const fleetAgents = agentsResponse?.agents.reduce((cache, agent) => {
+      cache.set(agent.id, { policy_id: agent.policy_id, policy_version: agent.policy_revision });
+      return cache;
+    }, new Map<string, FleetAgentCacheItem>());
+
+    const endpointPolicyCache = new Map<string, FullAgentPolicyInput>();
+    for (const policyInfo of fleetAgents.values()) {
+      if (policyInfo.policy_id !== null && policyInfo.policy_id !== undefined) {
```

Comment:
Just a personal preference here, but might make it a bit easier to read intent if we store the boolean logic in a variable
```
const shouldCachePolicy = 
	policyInfo.policy_id !== null &&
	policyInfo.policy_id !== undefined &&
	!endpointPolicyCache.has(policyInfo.policy_id)
```

Reply:
Yeah, that is slick. Thanks for the feedback!